File: backend/app/schemas/subscription.py
from datetime import datetime
from decimal import Decimal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
)

from app.models.owner_subscription import (
    BillingInterval,
    SubscriptionStatus,
)


class SubscriptionPlanRead(BaseModel):
    id: int
    code: str
    name: str

    monthly_price: Decimal
    yearly_price: Decimal
    max_properties: int | None

    sort_order: int
    is_active: bool

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


class SubscriptionPlanCreate(BaseModel):
    code: str = Field(
        min_length=2,
        max_length=50,
        pattern=r"^[A-Za-z0-9_-]+$",
    )

    name: str = Field(
        min_length=2,
        max_length=100,
    )

    monthly_price: Decimal = Field(
        ge=0,
        decimal_places=2,
    )

    yearly_price: Decimal = Field(
        ge=0,
        decimal_places=2,
    )

    max_properties: int | None = Field(
        default=None,
        ge=1,
    )

    sort_order: int = Field(
        default=0,
        ge=0,
    )

    is_active: bool = True

    @field_validator("code")
    @classmethod
    def normalize_code(
        cls,
        value: str,
    ) -> str:
        return (
            value.strip()
            .upper()
        )

    @field_validator("name")
    @classmethod
    def normalize_name(
        cls,
        value: str,
    ) -> str:
        return value.strip()


class SubscriptionPlanUpdate(BaseModel):
    name: str | None = Field(
        default=None,
        min_length=2,
        max_length=100,
    )

    monthly_price: Decimal | None = Field(
        default=None,
        ge=0,
        decimal_places=2,
    )

    yearly_price: Decimal | None = Field(
        default=None,
        ge=0,
        decimal_places=2,
    )

    max_properties: int | None = Field(
        default=None,
        ge=1,
    )

    sort_order: int | None = Field(
        default=None,
        ge=0,
    )

    is_active: bool | None = None

    @field_validator("name")
    @classmethod
    def normalize_name(
        cls,
        value: str | None,
    ) -> str | None:
        if value is None:
            return None

        return value.strip()


class OwnerSubscriptionRead(BaseModel):
    id: int
    owner_id: int
    status: SubscriptionStatus
    billing_interval: BillingInterval | None
    current_period_end: datetime | None
    cancel_at_period_end: bool

    property_count: int
    effective_max_properties: int | None

    plan: SubscriptionPlanRead

File: frontend/src/types/subscription.ts
export type SubscriptionStatus =
  | "FREE"
  | "ACTIVE"
  | "INCOMPLETE"
  | "PAST_DUE"
  | "CANCELED";


export type BillingInterval =
  | "MONTHLY"
  | "YEARLY";


export type SubscriptionPaymentStatus =
  | "PAID"
  | "FAILED";


export interface SubscriptionPlan {
  id: number;
  code: string;
  name: string;

  monthly_price:
    | string
    | number;

  yearly_price:
    | string
    | number;

  max_properties:
    | number
    | null;

  sort_order: number;
  is_active: boolean;

  created_at: string;
  updated_at: string;
}


export interface OwnerSubscription {
  id: number;
  owner_id: number;

  status:
    SubscriptionStatus;

  billing_interval:
    | BillingInterval
    | null;

  current_period_end:
    | string
    | null;

  cancel_at_period_end:
    boolean;

  property_count: number;

  effective_max_properties:
    | number
    | null;

  plan:
    SubscriptionPlan;
}


export interface SubscriptionPaymentItem {
  id: number;
  plan_id: number;
  plan_code: string;
  plan_name: string;

  stripe_invoice_id: string;

  amount:
    | string
    | number;

  currency: string;

  status:
    SubscriptionPaymentStatus;

  created_at: string;
}


export interface PageMeta {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}


export interface AdminSubscriptionSummary {
  total: number;
  free: number;
  active_paid: number;
  incomplete: number;
  past_due: number;
  canceled: number;
}


export interface AdminSubscriptionItem {
  owner_id: number;
  owner_email: string;

  owner_first_name:
    | string
    | null;

  owner_last_name:
    | string
    | null;

  owner_is_active: boolean;

  plan_id: number;
  plan_code: string;
  plan_name: string;

  status:
    SubscriptionStatus;

  billing_interval:
    | BillingInterval
    | null;

  property_count: number;

  max_properties:
    | number
    | null;

  current_period_end:
    | string
    | null;

  cancel_at_period_end:
    boolean;
}


export interface AdminSubscriptionPage {
  items:
    AdminSubscriptionItem[];

  meta: PageMeta;
}


export interface AdminSubscriptionPaymentItem {
  id: number;

  owner_id: number;
  owner_email: string;

  owner_first_name:
    | string
    | null;

  owner_last_name:
    | string
    | null;

  plan_id: number;
  plan_code: string;
  plan_name: string;

  stripe_invoice_id: string;

  amount:
    | string
    | number;

  currency: string;

  status:
    SubscriptionPaymentStatus;

  created_at: string;
}


export interface AdminSubscriptionPaymentPage {
  items:
    AdminSubscriptionPaymentItem[];

  meta: PageMeta;
}
